### PR TITLE
gnrc_ndp: add support for address-less link-layers

### PR DESCRIPTION
**Pull request by authmillenon**
Allows NDP to handle link-layers without addresses. This is achieved by handling the updating of the neighbor cache outside the context of the SL2A handle function and giving that function a numerical return value that states the length of the link-layer address (or negative for an error)

### CODE REVIEW

**Comment by authmillenon** [comment]
For some reason I saw very strange behavior regarding state changes on native. When entering `ng_ndp_internal_set_state()` the state was always DELAY, independent of what it was called with.

**Comment by authmillenon** [comment]
Rebased to current master

**Comment by authmillenon** [comment]
Rebased to current master

**Comment by authmillenon** [comment]
Rebased to current master.

**Comment by cgundogan** [comment]
>For some reason I saw very strange behavior regarding state changes on native. When entering ng_ndp_internal_set_state() the state was always DELAY, independent of what it was called with.

Is this fixed, or do you still see this strange behavior?

**Comment by authmillenon** [comment]
> Is this fixed, or do you still see this strange behavior?

Neighbor is reachable and later stale, so it seems to be:

```
> ifconfig
ifconfig
Iface  7   HWaddr: ea:29:ba:c8:38:be 
           MTU:1500  
           inet6 addr: ff02::1/128  scope: local [multicast]
           inet6 addr: fe80::e829:baff:fec8:38be/64  scope: local
           inet6 addr: ff02::1:ffc8:38be/128  scope: local [multicast]
           
> ncache
ncache
IPv6 address                    if  L2 address                state       type
------------------------------------------------------------------------------
fe80::58ff:38ff:fead:9beb        7  5a:ff:38:ad:9b:eb         REACHABLE   -
> ncache
ncache
IPv6 address                    if  L2 address                state       type
------------------------------------------------------------------------------
fe80::58ff:38ff:fead:9beb        7  5a:ff:38:ad:9b:eb         STALE       -
```

**Comment by authmillenon** [comment]
Got some fixes over from #3049 that were supposed to be in here.

**Comment by authmillenon** [comment]
Rebased tu current master

**Comment by authmillenon** [comment]
Since this PR is the jump-of-point for most of my extensions to NDP a quick test of this (both on a address-less link-layer like SLIP and on a link-layer with addresses) is needed. We also can consider to just test link-layers with addresses, confirm that it breaks nothing there (as I did in https://github.com/RIOT-OS/RIOT/pull/3628#issuecomment-134726381) and merge it regardless if we have no slip-capable devices at hand (as I do).

**Comment by OlegHahm** [comment]
Can you describe how to test the two scenarios? Which platforms fulfill the criteria?

**Comment by authmillenon** [comment]
> Can you describe how to test the two scenarios? Which platforms fulfill the criteria?

* Scenario 1 (link-layer with addresses):
 - platforms: any with an IEEE 802.15.4 or Ethernet device
 - ping other platform (or Linux host for native)
 - check neighbor cache:
```
ncache
IPv6 address                    if  L2 address                state       type
------------------------------------------------------------------------------
<ip_addr>                     <if>  <l2_addr>                 <state>     -
```

* Scenario 2 (link-layer without addresses [SLIP]):
 - platforms: any platform with more than one UART (and periph/uart support, please consult `periph_conf.h` of the board)
 - connect board with host over additional UART
 - use dist/tools/tunslip to open a TUN interface for SLIP on that UART (or connect two boards via UART)
 - compile with slip and `-DGNRC_NETIF_NUMOF=2` (depending on the app you might need to copy the `slip_params.h` from `tests/slip` to your app's directory and set `SLIP_UART` and `SLIP_BAUDRATE` to the configuration of your UART).
 - set-up IP and default route for SLIP interface
 - ping other platform (or host)
 - check neighbor cache
```
ncache
IPv6 address                    if  L2 address                state       type
------------------------------------------------------------------------------
<ip_addr>                     <if>                            <state>     -
```

**Comment by OlegHahm** [comment]
I still don't know how to test scenario 2. I've chosen a STM32F3Discovery board and have two UARTs connected. Now I compile `tests/slip` with `SLIP_UART=0 CFLAGS=-DGNRC_NETIF_NUMOF=2 BOARD=stm32f3discovery make`. Is this correct and how do I set up the IP and default route? I'm not even seeing an interface by calling `ifconfig` in this application.

**Comment by authmillenon** [comment]
You need to set `SLIP_UART=UART_1` or whatever UART you are using. UART_0 is used by the shell already, so you can't use that (at least as long until someone™ adds support for the multiplexer if `tunslip` ;-)). For `tests/slip` you don't need an additional interface so you don't need to set GNRC_NETIF_NUMOF, but you need gnrc_ipv6_default IIRC.

**Comment by authmillenon** [comment]
Use `ifconfig` to add an IPv6 address an `fibroute` to set the default route as usual.

**Comment by authmillenon** [comment]
Of course you need gnrc_ipv6_router_default for fib

**Comment by OlegHahm** [comment]
Apparently, I don't have an interface to add the IP address, too.

File: sys/net/gnrc/network_layer/ndp/gnrc_ndp.c [hunk 1]
```diff
@@ -70,41 +89,34 @@ void gnrc_ndp_nbr_sol_handle(kernel_pid_t iface, gnrc_pktsnip_t *pkt,
...
-
     while (sicmpv6_size > 0) {
         ndp_opt_t *opt = (ndp_opt_t *)(buf + opt_offset);
-
         switch (opt->type) {
             case NDP_OPT_SL2A:
-                if (!gnrc_ndp_internal_sl2a_opt_handle(iface, pkt, ipv6, nbr_sol->type, opt)) {
+                if ((l2src_len = gnrc_ndp_internal_sl2a_opt_handle(pkt, ipv6, nbr_sol->type, opt,
+                                                                   l2src)) < 0) {
+                    /* -ENOTSUP can not happen */
```
Comment (OlegHahm) [inline sys/net/gnrc/network_layer/ndp/gnrc_ndp.c]: Why not?

Comment (OlegHahm) [inline sys/net/gnrc/network_layer/ndp/gnrc_ndp.c]: (Please comment in code, not here.)

File: sys/net/gnrc/network_layer/ndp/gnrc_ndp.c [hunk 2]
```diff
@@ -248,11 +218,53 @@ void gnrc_ndp_nbr_adv_handle(kernel_pid_t iface, gnrc_pktsnip_t *pkt,
...
+                else if (l2tgt_changed) {
+                    gnrc_ndp_internal_set_state(nc_entry, GNRC_IPV6_NC_STATE_STALE);
+                }
+
+                if (nbr_adv->flags & NDP_NBR_ADV_FLAGS_R) {
+                    nc_entry->flags |= GNRC_IPV6_NC_IS_ROUTER;
+                }
+                else {
+                    nc_entry->flags &= ~GNRC_IPV6_NC_IS_ROUTER;
+                    /* TODO: update FIB */
```
Comment (OlegHahm) [inline sys/net/gnrc/network_layer/ndp/gnrc_ndp.c]: What has to be done for the FIB here?

File: sys/net/gnrc/network_layer/ndp/gnrc_ndp.c [hunk 3]
```diff
@@ -248,11 +218,53 @@ void gnrc_ndp_nbr_adv_handle(kernel_pid_t iface, gnrc_pktsnip_t *pkt,
...
+                }
+            }
+            else if (l2tgt_changed &&
+                     gnrc_ipv6_nc_get_state(nc_entry) == GNRC_IPV6_NC_STATE_REACHABLE) {
+                gnrc_ndp_internal_set_state(nc_entry, GNRC_IPV6_NC_STATE_STALE);
             }
         }
     }
```
Comment (OlegHahm) [inline sys/net/gnrc/network_layer/ndp/gnrc_ndp.c]: this function could also benefit from being split up, but that's a different topic.

**Comment by authmillenon** [comment]
> Apparently, I don't have an interface to add the IP address, too.

Did you clean and recompiled with `USEMODULE=gnrc_ipv6_router_default`?

**Comment by OlegHahm** [comment]
> Did you clean and recompiled with USEMODULE=gnrc_ipv6_router_default?

Yes.

**Comment by authmillenon** [comment]
Curious… sadly I don't have any board to test this… so it would be really nice of you to debug this... If you don't have time for this we could just use the second option in https://github.com/RIOT-OS/RIOT/pull/3628#issuecomment-136654708 and I'll provide a fixing PR in the feature freeze stage (when I get a board next week).

**Comment by authmillenon** [comment]
(or when `periph/uart` is working for native ^^)

**Comment by OlegHahm** [comment]
Just checking, that I got everything right.

My modified Makefile for `tests/slip` looks like:
```make

APPLICATION = driver_slip
include ../Makefile.tests_common

BOARD_INSUFFICIENT_MEMORY := stm32f0discovery

USEMODULE += auto_init_gnrc_netif
USEMODULE += gnrc_netif
USEMODULE += gnrc
USEMODULE += gnrc_pktdump
USEMODULE += gnrc_slip
USEMODULE += shell
USEMODULE += shell_commands
USEMODULE += gnrc_ipv6_router_default
USEMODULE += fib 
USEMODULE += ps

CFLAGS += -DDEVELHELP

CFLAGS += -I$(CURDIR)

ifneq (,$(SLIP_UART))
  CFLAGS += -DSLIP_UART=$(SLIP_UART)
else
  # set default
  CFLAGS += -DSLIP_UART="(UART_NUMOF-1)"
endif
ifneq (,$(SLIP_BAUDRATE))
  CFLAGS += -DSLIP_BAUDRATE=$(SLIP_BAUDRATE)
else
  # set default
  CFLAGS += -DSLIP_BAUDRATE=115200
endif

include $(RIOTBASE)/Makefile.include
```
and I compile and flash with
`SLIP_UART=UART_1 CFLAGS=-DGNRC_NETIF_NUMOF=2 BOARD=stm32f3discovery make -B clean all flash`.

**Comment by authmillenon** [comment]
It used to work -.- (even with out IPv6…)

**Comment by OlegHahm** [comment]
I think I got it.

**Comment by OlegHahm** [comment]
See #3757.

**Comment by OlegHahm** [comment]
But pinging still doesn't work.

**Comment by authmillenon** [comment]
Gorram! :(

**Comment by OlegHahm** [comment]
I'm on it.

**Comment by OlegHahm** [comment]
Might be a PEBCAK - I'm investigating.

**Comment by OlegHahm** [comment]
oops, got hardfault...

**Comment by OlegHahm** [comment]
For some reason neither RIOT nor Linux reacts to the Neighbor Solicitation.

**Comment by OlegHahm** [comment]
Hm, no, only Linux isn't responding to RIOT's neighbor solicitation. Maybe this is because of the tun interface?

**Comment by authmillenon** [comment]
Mh… can you maybe inject them with `scapy` for now? I'll try to figure out, what is wrong.

**Comment by OlegHahm** [comment]
How could I add the ncache entry manually in RIOT?

**Comment by authmillenon** [comment]
`ncache add <if> <ip-addr> <ll-addr>`

**Comment by authmillenon** [comment]
oh… the command can't handle address-less link-layers currently ^^ lemme fix that

**Comment by OlegHahm** [comment]
Gotta go now. However, in the worst case I'm willing to ACK this even if Linux-interaction over slip is not working for now.

**Comment by OlegHahm** [comment]
Just thinking about it: isn't it quite pointless to send out neighbor solicitations on an addressless link-layer anyway?

**Comment by authmillenon** [comment]
Why, for NUD it is still required.

**Comment by cgundogan** [comment]
Well, neighbor solicitations are also used to verify if a neighbor is reachable or not.

**Comment by authmillenon** [comment]
Adapted `ncache` shell command

**Comment by OlegHahm** [comment]
How do you know if a neighbor is reachable if you cannot address it?

**Comment by authmillenon** [comment]
You send a neighbor solicitation (either to the IPv6 address of the node if you already know the host [from previous communiction] or to the solicited-nodes multicast address of the IPv6 address if you don't know it) with the IPv6 address of the node in it's target field. If you get a neighbor advertisement with the solicited flag set and the target address being set to the IPv6 address you tested the reachability for, you know the neighbor is reachable.

**Comment by OlegHahm** [comment]
Hm, okay, but still the Linux side doesn't seem convinced that answering a neighbor solicitation for a tun interface. What about using a tap interface on the Linux side? Wouldn't this make things easier and remove the need for any exception case on the RIOT side?

**Comment by authmillenon** [comment]
a) a TAP interface is an Ethernet interface and SLIP is an other link-layer. So I don't see how you would use a TAP, at least not with some effort in `tunslip`. Are you sure it is because the TUN interface? (Linux' own [IPv6 incapable] SLIP tool `slattach` also creates a TUN btw).
b) this is definitely out of scope for this PR (let's discuss this in an issue or something like that).

**Comment by OlegHahm** [comment]
Ok, we don't have the time right now to debug this problem. The rest seems working (though I'm not happy with the rather big changes to "normal" ND), so let's move on. Please squash.

**Comment by authmillenon** [comment]
Squashe

**Comment by authmillenon** [comment]
Fixed error in shell and squashed immediately

**Comment by OlegHahm** [comment]
If I add the Linux host manually to the neighbor cache by calling something like `ncache add 5 fe80::1` the ping to `fe80::1` is sent with the unspecified address as source address...

**Comment by authmillenon** [comment]
Have you set a source address on the interface?

**Comment by OlegHahm** [comment]
lol - I did, but apparently rebooted afterwards. Now it's working.

**Pull request merged**